### PR TITLE
Review

### PR DESCRIPTION
**Pull request by hawx**
I thought this might be the easiest way to review the code, and should make a good start for jump starting discussions around standards/choices/opinions.

The commits should explain and show the changes I would make, the key things were:
- Label struct fields: makes the test cases easier to understand, and you can omit values that aren't needed
- Move the code under `internal`: not something I personally do, but I think if we broadly follow <https://github.com/golang-standards/project-layout> it will make life easier
- Merged some code from the previous packages to reduce the exported area
- Moved the module to github.com/ministryofjustice/opg-file-service
- Fixed warnings from `go vet`
- Moved all config to `main`

I also ran `go fmt` over everything, it would probably be a good idea to have that run in a pre-commit hook.

Another good tool to look at is <https://golangci-lint.run/>, which is still reporting 1 error in the actual code, but I can't be bothered fixing it :smile:

### CODE REVIEW

**Comment by TomTucka** [comment]
This makes sense and I like the structure. We can also run go fmt/golangci-lint in our pipeline and fail a PR to ensure a standard across the whole project.

I did have another read of https://github.com/golang-standards/project-layout and I now like some of the concepts it talks about 🙂

File: internal/handlers/zip.go [hunk 1]
```diff
@@ -1,37 +1,38 @@
...
-	"opg-file-service/middleware"
-	"opg-file-service/session"
-	"opg-file-service/zipper"
 	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/ministryofjustice/opg-file-service/internal/storage"
 )
 
+type Repository interface {
```
Comment (williamfalconeruk) [inline internal/handlers/zip.go:12]: just a question really - is it a best practice in go to have multiple interfaces defined in the same file? trying to understand coming from a c# background, and the rationale behind this having seen this in other go code too...

File: internal/storage/entry.go [hunk 2]
```diff
@@ -54,11 +54,6 @@ func (entry Entry) Validate() (bool, []error) {
 }
 
 func (entry Entry) DeDupe() {
-	type fileCounter struct {
```
Comment (williamfalconeruk) [inline internal/storage/entry.go:57]: another one -is it not necessary to declare the type? I noticed further down that there was a reference to this slice on line 79...again my ignorance of the language, probably...

File: internal/handlers/zip.go (hunk 1 above)
Comment (hawx) [reply internal/handlers/zip.go:12]: Yeah, generally the idea is to put things where they are used. This interface is only used in `ZipHandler` so it makes sense to keep it close by. Take a look at [this example](https://github.com/Evertras/go-interface-examples/blob/master/local-interfaces/handlers/user.go), and [the README](https://github.com/Evertras/go-interface-examples/tree/master/local-interfaces) for an explanation.

File: internal/storage/entry.go (hunk 2 above)
Comment (hawx) [reply internal/storage/entry.go:57]: There is already a definition for this type [further up the file](https://github.com/ministryofjustice/opg-file-service/pull/20/files/a2931d775da8870cf363dd898cfa1b7559667f89#diff-d822bdefef4fd1e6230d6db2174ad28fR16). But also it is kind of weird to define types _inside_ functions, they should live at the top-level.

Comment (williamfalconeruk) [reply internal/storage/entry.go:57]: sorry missed that one! agree not the right place, at least from a structural point of view.

**Comment by williamfalconeruk** [comment]
agree if we can keep to the suggested structure in the go project layout repo, and let the tools do some of the work to keep the code clean -  I think it's good - seems easy enough to read and navigate.
I think it would be good if we could get an example web app up at some point, just to see how that would work in practice.

**Comment by acsauk-mojd** [comment]
The project layout looks good and I forgot how much I love the structure of the table-driven tests - so easy to expand with more tests cases. This would be a good example project to base other micro-services on but agree with Billy that the same exercise for a full-on app would be useful. 

If I had to nominate a service, Serve-OPG is ripe for a re-write and pretty tiny in scope 😁

File: internal/handlers/zip.go (hunk 1 above)
Comment (cooperaj) [reply internal/handlers/zip.go:12]: > Imagine if instead there was some IDb interface from a traitorous C# coder that contained 50 different function signatures, but all leaderboard ever needed was that one.

I laughed

Comment (williamfalconeruk) [reply internal/handlers/zip.go:12]: > > Imagine if instead there was some IDb interface from a traitorous C# coder that contained 50 different function signatures, but all leaderboard ever needed was that one.
> 
> I laughed

remember I _was_ that traitorous c# coder 🤣

**Pull request closed without merging**